### PR TITLE
fix(core): Implement Boundary-First Tag Resolution

### PR DESCRIPTION
**Pull request by mm-parthy**
This PR implements the "Boundary-First Tag Resolution" pattern outlined in the [design document](https://gist.github.com/mm-parthy/e32a0aa1cb16b62d9ce6e80bfb436998).

The core problem was that the task `tag` was being resolved inconsistently across the application, which created multiple sources of truth and led to race conditions in the long-running MCP server and unpredictable behavior in the CLI.

This change refactors the entire stack to enforce a single, deterministic flow:
1.  **Boundary Resolves**: The effective tag is resolved **once** at the boundary layer (CLI command handler or MCP tool).
2.  **Context Carries**: A `context { projectRoot, tag }` object is passed explicitly through all function calls.
3.  **IO Obeys**: The low-level `readJSON` and `writeJSON` helpers now require an explicit `tag` and no longer perform their own discovery. (Currently they still do, final refactor for that needs some additional work )

This eliminates race conditions, makes state changes predictable, and simplifies debugging and testing.

## Type of Change

- [x] Bug fix (fixes race conditions and unpredictable state)
- [ ] New feature
- [x] Breaking change (internal function signatures have changed to require explicit context)
- [ ] Documentation update

## Testing

- [x] All existing tests have been updated to pass an explicit `{ tag }` context.
- [x] New tests have been added to verify that omitting the tag now correctly throws an error.
- [x] CLI and MCP tool test suites have been updated to reflect the new calling convention.
- [x] I have tested this locally.

## Changeset

- [x] I have created a changeset for this change.

## Additional Notes

This is a large but critical refactoring that touches many files across the CLI, MCP server, and core modules as detailed in the "Work Items" section of the design doc. The changes make our multi-context task management significantly more robust.

closes: #931 #857 #904 #843 

## Summary by CodeRabbit

* **New Features**
  * Added consistent, deterministic tag handling across all user-facing tools and commands, enabling explicit tag selection for operations such as adding, updating, moving, and removing tasks or dependencies.
  * Tools and CLI commands now accept an optional tag parameter, allowing users to specify the tag context for their actions.
  * Enhanced complexity report and file path handling to support tag-specific outputs.
  * Introduced a boundary-first tag resolution approach to ensure predictable tag processing.
  * Added support for specifying custom complexity report paths in relevant commands.

* **Bug Fixes**
  * Resolved potential race conditions and inconsistencies in tag resolution by centralizing and standardizing tag handling.

* **Documentation**
  * Improved user-facing documentation and help messages to reflect new tag-related options.
  * Enhanced inline documentation for parameters and options in various commands and tools.

* **Tests**
  * Expanded and updated test coverage to ensure correct tag handling and boundary conditions across all supported operations.
  * Added new tests for tag management and task operations within tag contexts.

* **Refactor**
  * Unified tag and project root handling across the codebase for improved maintainability and reliability.
  * Removed fallback logic for tag resolution, requiring explicit tag context where appropriate.
  * Centralized complexity report path resolution and tag awareness within core components.

### CODE REVIEW

**Comment by mm-parthy** [comment]
Closes: #947

File: scripts/modules/ui.js [hunk 1]
```diff
@@ -2602,7 +2604,6 @@ async function displayMultipleTasksSummary(
 				choice.trim(),
 				complexityReportPath,
 				statusFilter,
-				tag,
```
Comment (Crunchyman-ralph) [inline scripts/modules/ui.js:2605]: why remove the tag in the display ?

Comment (mm-parthy) [reply scripts/modules/ui.js:2605]: Removed for consistency, 
it's passed via Context.

**Comment by mm-parthy** [comment]
Closes: #981

**Comment by Crunchyman-ralph** [comment]
@coderabbit review

**Comment by Crunchyman-ralph** [comment]
@mm-parthy please address all of code rabbit's issues, I'm experimenting with this at the moment, grab your favorite ai agent and have them run this command:
```
gh api \
  -H "Accept: application/vnd.github+json" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  /repos/eyaltoledano/claude-task-master/pulls/943/comments
```

Then tell them to take coderabbit's "prompt for ai agents" and apply all of them :D

File: scripts/modules/task-manager/move-task.js [hunk 2]
```diff
@@ -27,6 +21,7 @@ async function moveTask(
 	generateFiles = false,
 	options = {}
 ) {
+	const { projectRoot, tag } = options;
```
[the remark this answers is not in the document]
Comment (Crunchyman-ralph) [reply scripts/modules/task-manager/move-task.js:24]: I think thats too strict especially since we have a way of getting the projectRoot and tag in case its not defined.

Though I might be wrong, could you please analyze and let me know ? I trust your judgement

File: scripts/modules/ui.js [hunk 3]
```diff
@@ -1197,18 +1197,18 @@ async function displayNextTask(
...
 	tasksPath,
 	taskId,
 	complexityReportPath = null,
 	statusFilter = null,
-	tag = null,
 	context = {}
 ) {
-	// Extract projectRoot from context
-	const projectRoot = context.projectRoot || null;
+	const { projectRoot, tag } = context;
```
[the remark this answers is not in the document]
Comment (mm-parthy) [reply scripts/modules/ui.js:1211]: This is to strict as these are resolved at the boundry. If you still think it's needed, let me know!

File: scripts/modules/task-manager/list-tasks.js [hunk 4]
```diff
@@ -36,18 +37,18 @@ function listTasks(
...
 	try {
 		// Extract projectRoot from context if provided
-		const projectRoot = context.projectRoot || null;
 		const data = readJSON(tasksPath, projectRoot, tag); // Pass projectRoot to readJSON
 		if (!data || !data.tasks) {
 			throw new Error(`No valid tasks found in ${tasksPath}`);
 		}
 
 		// Add complexity scores to tasks if report exists
+		//TODO: add tag to readComplexityReport and use it here
```
[the remark this answers is not in the document]
Comment (mm-parthy) [reply scripts/modules/task-manager/list-tasks.js]: This TODO is Obsolete, the ReportPath is handled at Boundry
